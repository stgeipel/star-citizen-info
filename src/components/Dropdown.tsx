import * as React from 'react';
import clsx from 'clsx';

export interface Props {
  className?: string | Record<string, unknown>;
  buttonLabel: string | React.ReactElement;
}

const Dropdown: React.FC<Props> = ({ className, buttonLabel, children }) => {
  const node = React.useRef<HTMLDivElement>();
  const [showDropdown, setShowDropdown] = React.useState(false);

  const toggleDropdown = () => {
    setShowDropdown(!showDropdown);
  };

  const handleClickOutside = (e: MouseEvent) => {
    if (node?.current?.contains(e.target as Node)) {
      return;
    }
    setShowDropdown(false);
  };

  React.useEffect(() => {
    document.addEventListener('mousedown', handleClickOutside);
    return () => document.removeEventListener('mousedown', handleClickOutside);
  }, []);

  return (
    <div ref={node} className={clsx('relative inline-block', className)}>
      <button
        className={clsx('p-2 rounded-full', {
          'hover:bg-primary hover:bg-opacity-25': !showDropdown,
          'bg-primary text-white': showDropdown,
        })}
        onClick={toggleDropdown}
      >
        {buttonLabel}
      </button>
      {showDropdown && (
        <div className="absolute right-0 mt-1 bg-white dark:bg-gray-800 border dark:border-gray-700 rounded shadow-md p-2 space-y-2">
          {children}
        </div>
      )}
    </div>
  );
};

export default Dropdown;
