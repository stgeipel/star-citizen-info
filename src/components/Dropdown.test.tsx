import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Dropdown from './Dropdown';

describe('Dropdown', () => {
  it('matches snapshot', () => {
    const wrapper = shallow(
      <Dropdown buttonLabel="menu">
        <ul>
          <li>hello</li>
          <li>world</li>
        </ul>
      </Dropdown>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
