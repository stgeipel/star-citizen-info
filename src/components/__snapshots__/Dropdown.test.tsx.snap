// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown matches snapshot 1`] = `
<div
  className="relative inline-block"
>
  <button
    className="p-2 rounded-full hover:bg-primary hover:bg-opacity-25"
    onClick={[Function]}
  >
    menu
  </button>
</div>
`;
